src/touch_save_app.c

```diff
@@ -0,0 +1,514 @@
+    .colorMode = GX_BG_COLORMODE_16,
+    .screenBase = GX_BG_SCRBASE_0x6800,
+    .charBase = GX_BG_CHARBASE_0x00000,
+    .bgExtPltt = GX_BG_EXTPLTT_01,
+    .priority = 2,
+    .areaOver = GX_BG_AREAOVER_XLU,
+    .dummy = 0,
+    .mosaic = FALSE,
+};
+
+typedef BOOL (*TouchSaveAppStateFunc)(TouchSaveAppData*);
+static TouchSaveAppStateFunc sTouchSaveAppStateFuncs[TOUCHSAVEAPP_STATE_COUNT] = {
+	TouchSaveApp_DisplaySaveInformation,
+	TouchSaveApp_AskForSave,
+	TouchSaveApp_WaitForTextPrinter,
+	TouchSaveApp_GetSaveConfirmation,
+	TouchSaveApp_HandleSaveConfirmation,
+	TouchSaveApp_PrintOverwriteMessage,
+	TouchSaveApp_GetOverwriteConfirmation,
+	TouchSaveApp_HandleOverwriteConfirmation,
+	TouchSaveApp_PrintNotMySaveError,
+	TouchSaveApp_SaveFailed,
+	TouchSaveApp_PrintSavingMessage,
+	TouchSaveApp_SetupWaitingIcon,
+	TouchSaveApp_SaveGame,
+	TouchSaveApp_PrintSavedMessage,
+	TouchSaveApp_SaveSucceeded,
+	TouchSaveApp_CloseApp,
+};
+
+SysTask *ov30_0225D520(BgConfig *bgConfig, void* a1, FieldSystem *fieldSystem, void* a3) {
+    CreateHeap(HEAP_ID_3, HEAP_ID_8, 0x18000);
+
+    GXS_SetGraphicsMode(GX_BGMODE_0);
+    GX_SetBankForSubBG(GX_VRAM_SUB_BG_32_H);
+    GX_SetBankForSubOBJ(GX_VRAM_SUB_OBJ_16_I);
+    // FIXME: Make an inline function for this.
+    reg_GXS_DB_DISPCNT = (reg_GXS_DB_DISPCNT & ~(REG_GXS_DB_DISPCNT_EXOBJ_CH_MASK | REG_GXS_DB_DISPCNT_OBJMAP_CH_MASK)) | REG_GXS_DB_DISPCNT_OBJMAP_CH_MASK;
+
+    InitBgFromTemplate(bgConfig, GF_BG_LYR_SUB_0, &ov30_0225DC2C, GF_BG_TYPE_TEXT);
+    InitBgFromTemplate(bgConfig, GF_BG_LYR_SUB_1, &ov30_0225DC48, GF_BG_TYPE_TEXT);
+    InitBgFromTemplate(bgConfig, GF_BG_LYR_SUB_2, &ov30_0225DC64, GF_BG_TYPE_TEXT);
+    BG_ClearCharDataRange(GF_BG_LYR_SUB_2, 0x20, 0, HEAP_ID_4);
+    BgClearTilemapBufferAndCommit(bgConfig, GF_BG_LYR_SUB_2);
+    BgClearTilemapBufferAndCommit(bgConfig, GF_BG_LYR_SUB_1);
+
+    GXS_SetVisibleWnd(0);
+    GX_EngineBToggleLayers(GF_BG_LYR_SUB_0_F, GX_LAYER_TOGGLE_OFF);
+    GX_EngineBToggleLayers(GF_BG_LYR_SUB_1_F, GX_LAYER_TOGGLE_OFF);
+    GX_EngineBToggleLayers(GF_BG_LYR_SUB_2_F, GX_LAYER_TOGGLE_OFF);
+    GX_EngineBToggleLayers(16, GX_LAYER_TOGGLE_OFF);
+
+    SysTask *task = sub_02007200((SysTaskFunc)ov30_0225D700, sizeof(TouchSaveAppData), 10, HEAP_ID_8);
+    TouchSaveAppData *data = sub_0201F988(task);
+    data->task = task;
+    data->unk0 = 0;
+    data->bgConfig = bgConfig;
+    data->unk18 = a1;
+    data->fieldSystem = fieldSystem;
+    data->unk30 = 0;
+    data->options = Save_PlayerData_GetOptionsAddr(data->fieldSystem->saveData);
+    data->state = TOUCHSAVEAPP_STATE_DISPLAY_SAVE_INFORMATION;
+
+    TouchSaveApp_SetupGraphics(data);
+    TouchSaveApp_SetupText(data, data->fieldSystem->unkD2_0);
+
+    GX_EngineBToggleLayers(GF_BG_LYR_SUB_0_F, GX_LAYER_TOGGLE_ON);
+    GX_EngineBToggleLayers(GF_BG_LYR_SUB_1_F, GX_LAYER_TOGGLE_ON);
+    GX_EngineBToggleLayers(GF_BG_LYR_SUB_2_F, GX_LAYER_TOGGLE_ON);
+    GX_EngineBToggleLayers(16, GX_LAYER_TOGGLE_ON);
```

Comment:
constant

Reply:
To my knowledge, it doesn't exist, unless I'm pulling it from the wrong enum

Reply:
is it not BG LYR SUB 3 F? or is that 8?

Reply:
It's 8

Reply:
looking at the docs, the value of 16 correlates to the OBJ planemask, since the entire func is named on layers, and it's to do with planemasks, imo a significant refactor is needed, at the very least for this PR, a different enum (planemask enum) should be used for this func